### PR TITLE
Correct @flow types. Refactor some Parser code for stricter typing.

### PR DESCRIPTION
**Pull request by marcianx**
Toward #892.

### CODE REVIEW

File: src/defineFunction.js [hunk 1]
```diff
@@ -16,6 +16,12 @@ export type FunctionContext = {|
 |};
 
 // TODO: Enumerate all allowed output types.
+
+// TODO: The real type of `args` is `(?ParseNode)[]`, but doing that breaks
+//       compilation since some of these arguments are passed to `ordargument`
+//       below which requires a non-nullable argument.
```
Comment (kevinbarabash) [inline src/defineFunction.js]: I like this idea.  It should be pretty easy to do too.

File: src/Parser.js [hunk 2]
```diff
@@ -543,47 +543,49 @@ class Parser {
...
-                return baseGroup.result;
+    /**
+     * Same as parseFunction(), except that the base is provided, guaranteeing a
+     * non-nullable result.
+     *
+     * @param {ParseFuncOrArgument} baseGroup
+     * @return {ParseNode}
+     */
+    parseGivenFunction(baseGroup) {
+        if (baseGroup.isFunction) {
```
Comment (kevinbarabash) [inline src/Parser.js]: I feel like by storing `isFunction` on `ParseFuncOrArgument` it leads to unnecessary wrapping and unwrapping.  What if we create a function that given a `ParseNode` would tell us if it was a function?  With that, we should be able to get rid of `ParseFuncOrArgument` and just use `ParseNode` everywhere.  This is maybe something to consider for a future PR.

File: src/Parser.js [hunk 3]
```diff
@@ -543,47 +543,49 @@ class Parser {
...
-            } else {
-                return baseGroup.result;
+    /**
+     * Same as parseFunction(), except that the base is provided, guaranteeing a
+     * non-nullable result.
+     *
+     * @param {ParseFuncOrArgument} baseGroup
+     * @return {ParseNode}
+     */
+    parseGivenFunction(baseGroup) {
```
Comment (kevinbarabash) [inline src/Parser.js]: Good idea separate this from the `parseFunction` which might return `null`.

File: src/Parser.js (hunk 2 above)
Comment (marcianx) [reply src/Parser.js]: Yes, I intend to do more changes within `Parser.js` itself in subsequent PRs while the `ParseNode` refactor waits on other PRs to complete. `ParseFuncOrArgument` will be replaced with a union of 3 structs per our discussion in #892. We can see whether it makes sense to have a function for an is-function check or simply use `foo.type === "fn"` (or `"$"` or `"arg"`) or some such.

**Comment by kevinbarabash** [comment]
@marcianx thanks for keeping the changes small.

**Pull request merged**